src/sshfs_mount/sshfs_mount.cpp

```diff
@@ -27,36 +27,78 @@
-template <typename Callable>
-auto run_cmd(mp::SSHSession& session, std::string&& cmd, Callable&& error_handler)
+const std::string fuse_version_string{"FUSE library version"};
+const std::string ld_library_path_key{"LD_LIBRARY_PATH="};
+const std::string snap_path_key{"SNAP="};
+
+auto run_cmd(mp::SSHSession& session, std::string&& cmd)
 {
     auto ssh_process = session.exec(cmd);
     if (ssh_process.exit_code() != 0)
-        error_handler(ssh_process);
-    return ssh_process.read_std_output();
-}
+        throw std::runtime_error(ssh_process.read_std_error());
 
-auto run_cmd(mp::SSHSession& session, std::string&& cmd)
-{
-    auto error_handler = [](mp::SSHProcess& proc) { throw std::runtime_error(proc.read_std_error()); };
-    return run_cmd(session, std::forward<std::string>(cmd), error_handler);
+    return ssh_process.read_std_output() + ssh_process.read_std_error();
 }
 
-void check_sshfs_exists(mp::SSHSession& session)
+auto get_sshfs_exec_and_options(mp::SSHSession& session)
 {
-    auto error_handler = [](mp::SSHProcess& proc) {
-        mpl::log(mpl::Level::warning, category,
-                 fmt::format("Unable to determine if 'sshfs' is installed: {}", proc.read_std_error()));
-        throw mp::SSHFSMissingError();
-    };
-
-    run_cmd(session, "which sshfs", error_handler);
+    std::string sshfs_exec;
+
+    try
+    {
+        // Prefer to use Snap package version first
+        std::istringstream sshfs_env{run_cmd(session, "sudo multipass-sshfs.env")};
+
+        auto ld_library_path = mp::utils::match_line_for(sshfs_env, ld_library_path_key);
+        auto snap_path = mp::utils::match_line_for(sshfs_env, snap_path_key);
+        snap_path = snap_path.substr(snap_path_key.length());
+
+        sshfs_exec = fmt::format("env {} {}/bin/sshfs", ld_library_path, snap_path);
+    }
+    catch (const std::exception& e)
+    {
+        mpl::log(mpl::Level::debug, category, fmt::format("'multipass-sshfs' snap package is not installed: {}", e.what()));
+
+        // Fallback to looking for distro version if snap is not found
+        try
+        {
+            sshfs_exec = run_cmd(session, "sudo which sshfs");
+        }
+        catch (const std::exception& e)
+        {
+            mpl::log(mpl::Level::warning, category,
+                     fmt::format("Unable to determine if 'sshfs' is installed: {}", e.what()));
+            throw mp::SSHFSMissingError();
+        }
+    }
+
+    sshfs_exec = mp::utils::trim_end(sshfs_exec);
+
+    std::istringstream version_info{run_cmd(session, fmt::format("sudo {} -V", sshfs_exec))};
+
+    sshfs_exec += " -o slave -o transform_symlinks -o allow_other";
+
+    auto fuse_version_line = mp::utils::match_line_for(version_info, fuse_version_string);
+    if (!fuse_version_line.empty())
+    {
+        // split on the fuse_version_string along with 0 or more colons
+        auto tokens = mp::utils::split(fuse_version_line, fmt::format("{}:* ", fuse_version_string));
```

Comment:
I wonder if we should be strict about it? I.e. match `^{}:? (0-9\.)+$`?

Reply:
I don't think it's necessary since I think we can trust the maintainers of FUSE to put in a correct version number here.  But if we do want to be very strict, then I think we need to account for more than a single digit followed by a `.` like `3.12.15` or something.

Also, `semver` should throw if the version is not something it can understand.

Reply:
OK you're right, semver will tell us if it's good. I would still only accept zero-or-one colon:
```suggestion
        auto tokens = mp::utils::split(fuse_version_line, fmt::format("{}:? ", fuse_version_string));
```